opendevin/controller/agent_controller.py

```diff
@@ -92,50 +103,127 @@ def __init__(
         self.max_iterations = max_iterations
         self.max_chars = max_chars
         self.workdir = workdir
-        self.command_manager = CommandManager(
-            self.id, workdir, container_image)
+        self.command_manager = CommandManager(self.id, workdir, container_image)
         self.callbacks = callbacks
 
     def update_state_for_step(self, i):
+        if self.state is None:
+            raise ValueError('State is not initialized')
         self.state.iteration = i
         self.state.background_commands_obs = self.command_manager.get_background_obs()
 
     def update_state_after_step(self):
+        if self.state is None:
+            raise ValueError('State is not initialized')
         self.state.updated_info = []
 
     def add_history(self, action: Action, observation: Observation):
+        if self.state is None:
+            raise ValueError('State is not initialized')
         if not isinstance(action, Action):
             raise TypeError(
-                f'action must be an instance of Action, got {type(action).__name__} instead')
+                f'action must be an instance of Action, got {type(action).__name__} instead'
+            )
         if not isinstance(observation, Observation):
             raise TypeError(
-                f'observation must be an instance of Observation, got {type(observation).__name__} instead')
+                f'observation must be an instance of Observation, got {type(observation).__name__} instead'
+            )
         self.state.history.append((action, observation))
         self.state.updated_info.append((action, observation))
 
-    async def start_loop(self, task: str):
-        finished = False
-        plan = Plan(task)
-        self.state = State(plan)
-        for i in range(self.max_iterations):
+    async def _run(self):
+        if self.state is None:
+            raise ValueError('State is not initialized')
+
+        if self._task_state != TaskState.RUNNING:
+            raise ValueError('Task is not in running state')
+
+        for i in range(self._cur_step, self.max_iterations):
             try:
-                finished = await self.step(i)
+                self._finished = await self.step(i)
             except Exception as e:
                 logger.error('Error in loop', exc_info=True)
                 raise e
-            if finished:
-                break
-        if not finished:
+
+            if self._task_state == TaskState.FINISHED:
+                await self._stop_task(is_finished=True)  # type: ignore[unreachable]
+                return
```

Comment:
should this be `break` so we hit the log statement below?

Reply:
actually, i don't what you mean 😅

Reply:
`break` means "exit the loop". `return` means "exit the function"

since you're returning here, we never reach the log statement that's below the loop